WordPress/Sniffs/Commenting/HooksInlineDocsSniff.php

```diff
@@ -0,0 +1,236 @@
+
+use WordPressCS\WordPress\AbstractFunctionRestrictionsSniff;
+use PHP_CodeSniffer\Util\Tokens;
+
+/**
+ * Class HooksMustHaveDocblockSniff
+ *
+ * @package WPCS\WordPressCodingStandards
+ */
+class HooksInlineDocsSniff extends AbstractFunctionRestrictionsSniff {
+
+	/**
+	 * Array of WordPress hook execution functions.
+	 *
+	 * @var array WordPress hook execution function name => filter or action.
+	 */
+	protected $hook_functions = array(
+		'apply_filters'            => 'filter',
+		'apply_filters_ref_array'  => 'filter',
+		'apply_filters_deprecated' => 'filter',
+		'do_action'                => 'action',
+		'do_action_ref_array'      => 'action',
+		'do_action_deprecated'     => 'action',
+	);
+
+	/**
+	 * Array of allowed exceptional version numbers.
+	 *
+	 * By default, X.Y.Z version numbers are required. If there are any exceptions,
+	 * they can be passed in the ruleset XML file via:
+	 * <rule ref="WordPress.Commenting.HooksInlineDocs">
+	 *  <properties>
+	 *   <property name="allowed_extra_versions" type="array" value="0.71"/>
+	 *   <property name="allowed_extra_versions" type="array" value="MU (3.0.0)"/>
+	 * </rule>
+	 *
+	 * The key values are used to determine if a version is allowed outside of the X.Y.Z scheme. The value is not considered.
+	 *
+	 * @var array Array of allowed exceptional values.
+	 */
+	public $allowed_extra_versions = array();
+
+	/**
+	 * Groups of functions to restrict.
+	 *
+	 * Example: groups => array(
+	 *  'lambda' => array(
+	 *      'type'      => 'error' | 'warning',
+	 *      'message'   => 'Use anonymous functions instead please!',
+	 *      'functions' => array( 'file_get_contents', 'create_function' ),
+	 *  )
+	 * )
+	 *
+	 * @return array
+	 */
+	public function getGroups() {
+		return array(
+			'hooks' => array(
+				'functions' => array_keys( $this->hook_functions ),
+			),
+		);
+	}
+
+	/**
+	 * Process a matched token.
+	 *
+	 * @since 1.0.0 Logic split off from the `process_token()` method.
+	 *
+	 * @param int    $stack_ptr       The position of the current token in the stack.
+	 * @param string $group_name      The name of the group which was matched.
+	 * @param string $matched_content The token content (function name) which was matched.
+	 *
+	 * @return int|void Integer stack pointer to skip forward or void to continue
+	 *                  normal file processing.
+	 */
+	public function process_matched_token( $stack_ptr, $group_name, $matched_content ) {
+
+		if ( ! $this->verify_valid_match( $stack_ptr ) ) {
+			return;
+		}
+
+		$previous_comment = $this->return_previous_comment( $stack_ptr );
+
+		if ( false !== $previous_comment ) {
+			/*
+			 * Check to determine if there is a comment immediately preceding the function call.
+			 */
+			if ( ( $this->tokens[ $previous_comment ]['line'] + 1 ) !== $this->tokens[ $stack_ptr ]['line'] ) {
+				$this->phpcsFile->addError(
+					'The inline documentation for a hook must be on the line immediately before the function call.',
```

Comment:
As noted previously, this may not always be possible. Consider this example from Juliette:

```
/**
 * Hook docblock
 */
if ( $a === true
    && apply_filter('some_filter', $value) > 10
) {
    // Do something.
}
```

Likewise, this should also be valid IMO, since it matches the PSR-5 definition of associating a DocBlock with a Structural Element with an optional number of clear lines in between:

```
/**
 * Hook docblock
 */

$value = apply_filters( 'some_filter', $value );
```